sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/NestedColumnAliasing.scala

```diff
@@ -0,0 +1,193 @@
+      Project(
+        getNewProjectList(projectList, nestedFieldToAlias),
+        g.copy(child = replaceChildrenWithAliases(grandChild, attrToAliases)))
+
+    case Project(projectList, child) =>
+      Project(
+        getNewProjectList(projectList, nestedFieldToAlias),
+        replaceChildrenWithAliases(child, attrToAliases))
+  }
+
+  /**
+   * Return a replaced project list.
+   */
+  private def getNewProjectList(
+      projectList: Seq[NamedExpression],
+      nestedFieldToAlias: Map[GetStructField, Alias]): Seq[NamedExpression] = {
+    projectList.map(_.transform {
+      case f: GetStructField if nestedFieldToAlias.contains(f) =>
+        nestedFieldToAlias(f).toAttribute
+    }.asInstanceOf[NamedExpression])
+  }
+
+  /**
+   * Return a plan with new childen replaced with aliases.
+   */
+  private def replaceChildrenWithAliases(
+      plan: LogicalPlan,
+      attrToAliases: Map[ExprId, Seq[Alias]]): LogicalPlan = {
+    plan.withNewChildren(plan.children.map { plan =>
+      Project(plan.output.flatMap(a => attrToAliases.getOrElse(a.exprId, Seq(a))), plan)
+    })
+  }
+
+  /**
+   * Returns true for those operators that project can be pushed through.
+   */
+  private def canProjectPushThrough(plan: LogicalPlan) = plan match {
+    case _: GlobalLimit => true
+    case _: LocalLimit => true
+    case _: Repartition => true
+    case _: Sample => true
+    case _ => false
+  }
+
+  /**
+   * Similar to [[QueryPlan.references]], but this only returns all attributes
+   * that are explicitly referenced on the root levels in a [[LogicalPlan]].
+   */
+  private def getRootReferences(plan: LogicalPlan): AttributeSet = {
+    def helper(e: Expression): AttributeSet = e match {
+      case attr: AttributeReference => AttributeSet(attr)
+      case _: GetStructField => AttributeSet.empty
+      case es if es.children.nonEmpty => AttributeSet(es.children.flatMap(helper))
+      case _ => AttributeSet.empty
+    }
+    AttributeSet.fromAttributeSets(plan.expressions.map(helper))
+  }
+
+  /**
+   * Returns all the nested fields that are explicitly referenced as [[Expression]]
+   * in a [[LogicalPlan]]. Currently, we only support having [[GetStructField]] in the chain
+   * of the expressions. If the chain contains GetArrayStructFields, GetMapValue, or
+   * GetArrayItem, the nested field substitution will not be performed.
+   */
+  private def getNestedFieldReferences(plan: LogicalPlan): Seq[GetStructField] = {
+    def helper(e: Expression): Seq[GetStructField] = e match {
+      case f @ GetStructField(child, _, _) if isGetStructFieldOrAttr(child) => Seq(f)
+      case es if es.children.nonEmpty => es.children.flatMap(e => helper(e))
+      case _ => Seq.empty
+    }
+
+    def isGetStructFieldOrAttr(e: Expression): Boolean = e match {
+      case GetStructField(child, _, _) => isGetStructFieldOrAttr(child)
+      case _: AttributeReference => true
+      case _ => false
+    }
+
+    plan.expressions.flatMap(helper)
+  }
```

Comment:
I feel `getRootReferences` and `getNestedFieldReferences` are a bit complicated, so how about this ?
https://github.com/apache/spark/commit/bbf5a5718fc9ab889f2e630edb40a0d4377f3ddc